File: Python/20201203.py
import numpy
import time


def extend_to_right(slope_map, steps):
    length = len(slope_map)

    for i, row in enumerate(slope_map):
        slope_map[i] = row * (length * steps[1] - 1)
    return slope_map


def find_path(slope_map, start_position=(0, 0), steps=(1, 3)):
    tree_count = 0

    rows = list(range(start_position[0], len(slope_map), steps[0]))
    columns = list(range(start_position[1], len(
        slope_map)*steps[1], steps[1]))

    for row, column in zip(rows, columns):
        if slope_map[row][column] == "#":
            tree_count += 1

    return tree_count


if __name__ == "__main__":
    slope_map = open("../input/day3.txt", "r")
    slope_map = slope_map.readlines()
    slope_map = [line.rstrip() for line in slope_map]
    slope_map = extend_to_right(slope_map, (1, 7))
    start = time.time()
    print(f"Part 1: {find_path(slope_map)}")
    end = time.time()
    print(f"Part 1 took: {end-start} seconds\n")

    part2_slopes = ((1, 1), (1, 3), (1, 5), (1, 7), (2, 1))
    part2_counts = []
    start = time.time()
    for slope in part2_slopes:
        part2_counts.append(find_path(slope_map, steps=slope))
        print(f"\tFinished {slope}, count was: {part2_counts[-1]}")
    product = numpy.prod(part2_counts)
    print(f"Part 2: {product}")
    end = time.time()
    print(f"Part 2 took: {end - start} seconds")

File: Go/20201203.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"
)

func readFile(fileName string) []string {
	f, _ := os.Open(fileName)

	defer f.Close()

	scanner := bufio.NewScanner(f)

	lines := []string{}

	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	return lines
}

func calcTrees(slopeMap []string, steps [2]int) int {
	column := 0
	targetCol := 0
	rowLength := len(slopeMap[0])

	treeNum := 0

	for row := 0; row < len(slopeMap); row += steps[0] {
		if column >= rowLength {
			targetCol = column % rowLength
		} else {
			targetCol = column
		}

		currentRow := strings.Split(slopeMap[row], "")
		if currentRow[targetCol] == "#" {
			treeNum++
		}

		column += steps[1]
	}
	return treeNum
}

func main() {
	slopeMap := readFile("../input/day3.txt")
	steps := [2]int{1, 3}

	start := time.Now()
	part1Result := calcTrees(slopeMap, steps)
	fmt.Println("Part 1 result: ", part1Result, "Took: ", time.Since(start))

	part2Results := 1

	part2Steps := [5][2]int{
		{1, 1},
		{1, 3},
		{1, 5},
		{1, 7},
		{2, 1},
	}

	start = time.Now()
	for _, steps := range part2Steps {
		part2Results = part2Results * calcTrees(slopeMap, steps)
	}
	fmt.Println("Part 2 result: ", part2Results, "Took: ", time.Since(start))
}
